p2p/src/sync/mod.rs

```diff
@@ -173,40 +157,59 @@ where
         Ok(receiver)
     }
 
-    /// Registers the connected peer by creating a context for it.
-    ///
-    /// The `HeaderListRequest` message is sent to newly connected peers.
-    async fn register_peer(&mut self, peer: PeerId) -> Result<()> {
+    /// Starts a task for the new peer.
+    pub fn register_peer(&mut self, peer: PeerId) -> Result<()> {
         log::debug!("Register peer {peer} to sync manager");
 
-        self.request_headers(peer).await?;
-        match self.peers.insert(peer, PeerContext::new()) {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        self.peers
+            .insert(peer, sender)
             // This should never happen because a peer can only connect once.
-            Some(_) => Err(P2pError::PeerError(PeerError::PeerAlreadyExists)),
-            None => Ok(()),
-        }
+            .map(|_| Err::<(), _>(P2pError::PeerError(PeerError::PeerAlreadyExists)))
+            .transpose()?;
+
+        let peer_sender = self.peer_sender.clone();
+        let peer_manager_sender = self.peer_manager_sender.clone();
+        let chainstate_handle = self.chainstate_handle.clone();
+        let p2p_config = Arc::clone(&self.p2p_config);
+        let is_initial_block_download = Arc::clone(&self.is_initial_block_download);
+        tokio::spawn(async move {
+            let mut peer = Peer::<T>::new(
+                peer,
+                p2p_config,
+                chainstate_handle,
+                peer_manager_sender,
+                peer_sender,
+                receiver,
+                is_initial_block_download,
+            );
+            if let Err(e) = peer.run().await {
+                log::error!("Sync manager peer ({}) error: {e:?}", peer.id());
+            }
+        });
+
+        Ok(())
     }
 
-    /// Removes the state (`PeerContext`) of the given peer.
+    /// Stops the task of the given peer by closing the corresponding channel.
     fn unregister_peer(&mut self, peer: PeerId) {
         log::debug!("Unregister peer {peer} from sync manager");
 
-        // Remove the queued block responses associated with the disconnected peer.
-        self.blocks_queue.retain(|(p, _)| p != &peer);
-
-        if self.peers.remove(&peer).is_some() {
+        if self.peers.remove(&peer).is_none() {
             log::warn!("Unregistering unknown peer: {peer}");
```

Comment:
Another part where I think maybe an expect is appropriate?

I'm really not sure. But maybe we should start with that and see what happens. We can throw in a TODO to keep watching it.

Reply:
I think that assertion (expect) here will be appropriate here because the `unregister_peer` function should never be called twice, so it indicates an error elsewhere. I'm not so sure about the previous comment though.

Reply:
It's always this difficult balance between "willing to crash the node if something goes wrong", and "possibility to miss the presence of an error".

Things like this, if we decide to keep them and log them, should be labeled "CRITICAL" in the logs to be noticeable as this should "never happen".